### PR TITLE
Fix close exception and server-info EXPIRED_INTERVAL

### PR DESCRIPTION
**Pull request by zyxxoo**
also fix "The server with name xx already in cluster"

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java [hunk 1]
```diff
@@ -53,6 +53,10 @@
 
     private final EventHub eventHub;
 
+    static {
+        HugeGraphAuthProxy.setContext(Context.admin());
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java:57]: can remove setContext() call at line 101 from injectAuthGraph()

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 2]
```diff
@@ -122,5 +122,11 @@ public static void main(String[] args) throws Exception {
             server.stop();
             LOG.info("HugeGraphServer stopped");
         }, "hugegraph-server-shutdown"));
+        /*
+         * HugeFactory shutdown hook be invoked before server stop, so that
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:126]: HugeFactory.shutdown hook may be ...

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 3]
```diff
@@ -122,5 +122,11 @@ public static void main(String[] args) throws Exception {
...
+         * eventHub execute notify event on a shutdown executor, so here
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:127]: causes event-hub can't execute notification events for another shutdown executor such as gremling-stop-shutdown

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 4]
```diff
@@ -122,5 +122,11 @@ public static void main(String[] args) throws Exception {
...
+         * remove the HugeFactory shutdown hook.
+         */
+        HugeFactory.removeShutdownHook();
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:130]: also add graph.tx().close() to checkVersionOrExit()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/HugeFactory.java [hunk 5]
```diff
@@ -138,6 +143,9 @@ public static PropertiesConfiguration getRemoteConfig(URL url) {
      * @param timeout seconds
      */
     public static void shutdown(long timeout) {
+        if (!shutdown.compareAndSet(false, true)) {
+            return;
+        }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/HugeFactory.java:148]: add log LOG.info("HugeGraph shutdown");

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 6]
```diff
@@ -122,5 +122,11 @@ public static void main(String[] args) throws Exception {
             server.stop();
             LOG.info("HugeGraphServer stopped");
         }, "hugegraph-server-shutdown"));
+        /*
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:125]: we can let main thread wait for server shutdown
```java
diff --git a/hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java b/hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java
index 8d5d5d585..513de8d63 100644
--- a/hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java
+++ b/hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java
@@ -19,6 +19,8 @@
 
 package com.baidu.hugegraph.dist;
 
+import java.util.concurrent.CompletableFuture;
+
 import org.apache.tinkerpop.gremlin.server.GremlinServer;
 import org.slf4j.Logger;
 
@@ -117,10 +119,19 @@ public class HugeGraphServer {
         HugeGraphServer.register();
 
         HugeGraphServer server = new HugeGraphServer(args[0], args[1]);
+
+        CompletableFuture<?> future = new CompletableFuture<>();
         Runtime.getRuntime().addShutdownHook(new Thread(() -> {
             LOG.info("HugeGraphServer stopping");
-            server.stop();
-            LOG.info("HugeGraphServer stopped");
+            try {
+                server.stop();
+                LOG.info("HugeGraphServer stopped");
+                future.complete(null);
+            } catch (Throwable e) {
+                future.completeExceptionally(e);
+            }
         }, "hugegraph-server-shutdown"));
+
+        future.get();
     }
 }
```

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 7]
```diff
@@ -337,28 +337,33 @@ private void checkBackendVersionOrExit(HugeConfig config) {
...
-                if (this.requireAuthentication()) {
-                    String token = config.get(ServerOptions.AUTH_ADMIN_TOKEN);
-                    try {
-                        this.authenticator.initAdminUser(token);
-                    } catch (Exception e) {
-                        throw new BackendException(
-                                  "The backend store of '%s' can't " +
-                                  "initialize admin user", hugegraph.name());
+            try {
+                if (!hugegraph.backendStoreFeatures().supportsPersistence()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:341]: can move to a new method?

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 8]
```diff
@@ -117,16 +119,19 @@ public static void main(String[] args) throws Exception {
         HugeGraphServer.register();
 
         HugeGraphServer server = new HugeGraphServer(args[0], args[1]);
+        /*
+         * HugeFactory.shutdown hook may be invoked before server stop,
+         * causes event-hub can't execute notification events for another
+         * shutdown executor such as gremling-stop-shutdown
+         */
+        HugeFactory.removeShutdownHook();
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:127]: add blank lines for removeShutdownHook();

**Review by javeme — APPROVED** [review]

**Review by imbajin — APPROVED** [review]

**Pull request merged**